registry/handlers/manifests.go

```diff
@@ -131,24 +267,99 @@ func (imh *manifestHandler) GetManifest(w http.ResponseWriter, r *http.Request)
 			if mediaType == manifestlist.MediaTypeManifestList {
 				supportsManifestList = true
 			}
+			if mediaType == ocischema.MediaTypeManifest {
+				supportsOCISchema = true
+			}
+			if mediaType == manifestlist.MediaTypeOCIManifestList {
+				supportsOCIManifestList = true
+			}
+		}
+	}
+	supportsOCI := supportsOCISchema || supportsOCIManifestList
+
+	ctxu.GetLogger(imh).Debug("GetImageManifest")
+	manifests, err := imh.Repository.Manifests(imh)
+	if err != nil {
+		imh.Errors = append(imh.Errors, err)
+		return
+	}
+
+	var manifest distribution.Manifest
+	if imh.Tag != "" {
+		tags := imh.Repository.Tags(imh)
+		var desc distribution.Descriptor
+		if !supportsOCI {
+			desc, err = tags.Get(imh, imh.Tag)
+			if err != nil {
+				imh.Errors = append(imh.Errors, v2.ErrorCodeManifestUnknown.WithDetail(err))
+				return
+			}
+		} else {
+			desc, err = tags.Get(imh, imh.annotatedTag(supportsOCI))
+			if err != nil {
+				desc, err = tags.Get(imh, imh.annotatedTag(false))
+				if err != nil {
+					imh.Errors = append(imh.Errors, v2.ErrorCodeManifestUnknown.WithDetail(err))
+					return
+				}
+			}
 		}
+		imh.Digest = desc.Digest
+	}
+
+	if etagMatch(r, imh.Digest.String()) {
+		w.WriteHeader(http.StatusNotModified)
+		return
+	}
+
+	var options []distribution.ManifestServiceOption
+	if imh.Tag != "" {
+		options = append(options, distribution.WithTag(imh.annotatedTag(supportsOCI)))
```

Comment:
I think this shouldn't just pass in supportsOCI here - it needs to handle the case above where the code falls back from looking for the annotated tag to looking for an unannotated tag.